cohttp-lwt-unix/src/bytebuffer.ml

```diff
@@ -0,0 +1,105 @@
+module Bytes = BytesLabels
+open Lwt.Infix
+
+(* Bytebuffer is split into three regions using two separate indices that are used
+   to support read and write operations.
+   +--------------------+---------------------------+----------------------------+
+   | Consumed Bytes     | Bytes available to read   | Empty space for writing    |
+   +--------------------+---------------------------+----------------------------+
+   |     0 <=       pos_read         <=          pos_fill              <= capacity
+
+   Consumed Bytes: This is content that's already consumed via a get/read operation.
+   This space can be safely reclaimed.
+
+   Bytes available to read: This is the actual content that will be surfaced to users via
+   get/read operations on the bytebuffer.
+
+   Empty space for writing: This is space that will be filled by any set/write operations
+   on the bytebuffer.
+*)
+
+type t = {
+  mutable buf : Bytes.t;
+  mutable pos_read : int;
+  mutable pos_fill : int;
+}
+
+let create size =
+  let buf = Bytes.create size in
+  { buf; pos_read = 0; pos_fill = 0 }
+
+let unsafe_buf t = t.buf
+let pos t = t.pos_read
+
+let compact t =
+  if t.pos_read > 0 then (
+    let len = t.pos_fill - t.pos_read in
+    Bytes.blit ~src:t.buf ~dst:t.buf ~src_pos:t.pos_read ~dst_pos:0 ~len;
+    t.pos_read <- 0;
+    t.pos_fill <- len)
+
+let length t = t.pos_fill - t.pos_read
+
+let drop t len =
+  if len < 0 || len > length t then
+    invalid_arg "Bytebuffer.drop: Index out of bounds";
+  t.pos_read <- t.pos_read + len
+
+let refill t reader =
+  compact t;
+  Lwt_io.read_into reader t.buf t.pos_fill (Bytes.length t.buf - t.pos_fill)
```

Comment:
```suggestion
  Lwt.catch
    (fun () ->
      Lwt_io.read_into reader t.buf t.pos_fill (Bytes.length t.buf - t.pos_fill)
      >|= fun count ->
      if count > 0 then (
        t.pos_fill <- t.pos_fill + count;
        `Ok)
      else `Eof)
    (function Lwt_io.Channel_closed _ -> Lwt.return `Eof | exn -> raise exn)
```
or
```suggestion
  Lwt.catch
    (fun () ->
      Lwt_io.read_into reader t.buf t.pos_fill (Bytes.length t.buf - t.pos_fill))
    (function Lwt_io.Channel_closed _ -> Lwt.return 0 | exn -> raise exn)
  >|= fun count ->
  if count > 0 then (
    t.pos_fill <- t.pos_fill + count;
    `Ok)
      else `Eof
```
both fix it in my tests

Reply:
Thanks! I missed that the channels can be closed in the expert response flow!